File: src/main/java/io/github/xf8b/xf8bot/commands/administration/RemoveWarnCommand.java
/*
 * Copyright (c) 2020 xf8b.
 *
 * This file is part of xf8bot.
 *
 * xf8bot is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * xf8bot is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with xf8bot.  If not, see <https://www.gnu.org/licenses/>.
 */

package io.github.xf8b.xf8bot.commands.administration;

import com.mongodb.client.model.Filters;
import com.mongodb.reactivestreams.client.MongoCollection;
import discord4j.common.util.Snowflake;
import io.github.xf8b.xf8bot.api.commands.AbstractCommand;
import io.github.xf8b.xf8bot.api.commands.CommandFiredEvent;
import io.github.xf8b.xf8bot.api.commands.flags.StringFlag;
import io.github.xf8b.xf8bot.util.ParsingUtil;
import org.bson.Document;
import org.bson.conversions.Bson;
import org.jetbrains.annotations.NotNull;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

public class RemoveWarnCommand extends AbstractCommand {
    private static final StringFlag MEMBER = StringFlag.builder()
            .setShortName("m")
            .setLongName("member")
            .setNotRequired()
            .build();
    private static final StringFlag MEMBER_WHO_WARNED = StringFlag.builder()
            .setShortName("mww")
            .setLongName("memberWhoWarned")
            .setNotRequired()
            .build();
    private static final StringFlag REASON = StringFlag.builder()
            .setShortName("r")
            .setLongName("reason")
            .setNotRequired()
            .build();

    private static final StringFlag WARN_ID = StringFlag.builder()
            .setShortName("i")
            .setLongName("warnId")
            //TODO: fix
            .setValidityPredicate(s -> s.matches("\\b[0-9a-f]{8}\\b-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-\\b[0-9a-f]{12}\\b"))
            .setInvalidValueErrorMessageFunction($ -> "The warn ID must be a UUID!")
            .setNotRequired()
            .build();

    public RemoveWarnCommand() {
        super(AbstractCommand.builder()
                .setName("${prefix}removewarn")
                .setDescription("""
                        Removes the specified member's warns with the warnId and reason provided.
                        If the reason is all, all warns will be removed. The warnId is not needed.
                        If the warnId is all, all warns with the same reason will be removed.
                        """)
                .setCommandType(CommandType.ADMINISTRATION)
                .setAliases("${prefix}removewarns", "${prefix}rmwarn", "${prefix}rmwarns")
                .setMinimumAmountOfArgs(2)
                .setFlags(MEMBER, MEMBER_WHO_WARNED, REASON, WARN_ID)
                .setAdministratorLevelRequired(1));
    }

    @NotNull
    @Override
    public Mono<Void> onCommandFired(@NotNull CommandFiredEvent event) {
        Mono<Snowflake> userId = Mono.justOrEmpty(event.getValueOfFlag(MEMBER))
                .flatMap(it -> ParsingUtil.parseUserId(event.getGuild(), it)
                        .map(Snowflake::of)
                        .switchIfEmpty(event.getChannel().flatMap(it1 -> it1.createMessage("The member does not exist!"))
                                .then()
                                .cast(Snowflake.class)));
        Mono<Snowflake> memberWhoWarnedId = Mono.justOrEmpty(event.getValueOfFlag(MEMBER_WHO_WARNED))
                .flatMap(it -> ParsingUtil.parseUserId(event.getGuild(), it)
                        .map(Snowflake::of)
                        .switchIfEmpty(event.getChannel().flatMap(it1 -> it1.createMessage("The member who warned does not exist!"))
                                .then()
                                .cast(Snowflake.class)));
        String reason = event.getValueOfFlag(REASON).orElse(null);
        String warnId = event.getValueOfFlag(WARN_ID).orElse(null);
        MongoCollection<Document> mongoCollection = event.getXf8bot()
                .getMongoDatabase()
                .getCollection("warns");
        Mono<Bson> filter = userId.map(it -> Filters.and(
                Filters.eq("guildId", event.getGuildId().get().asLong()),
                Filters.eq("userId", it.asLong())
        )).defaultIfEmpty(Filters.eq("guildId", event.getGuildId().get().asLong()));
        Flux<Document> warnDocuments = filter.flatMapMany(it -> Flux.from(mongoCollection.find(it)));
        return Mono.zip(
                Mono.just(warnId == null),
                memberWhoWarnedId.flux().count().map(it -> it == 0L),
                userId.flux().count().map(it -> it == 0L),
                Mono.just(reason == null)
        ).filter(it -> !it.getT1() && !it.getT2() && !it.getT3() && !it.getT4())
                .flatMap($ -> {
                    if (reason != null && reason.equalsIgnoreCase("all")) {
                        return userId.flatMap(it -> warnDocuments.flatMap(document -> Mono.from(mongoCollection.deleteOne(document)))
                                .flatMap($1 -> event.getGuild().flatMap(it1 -> it1.getMemberById(it)))
                                .flatMap(member -> event.getChannel().flatMap(it1 -> it1.createMessage("Successfully removed warn(s) for " + member.getDisplayName() + ".")))
                                .switchIfEmpty(event.getChannel().flatMap(it1 -> it1.createMessage("Cannot remove all warns without a user!")))
                                .then());
                    } else {
                        return warnDocuments
                                .filter(document -> {
                                    if (warnId != null) {
                                        return document.getString("warnId").equals(warnId);
                                    } else {
                                        return true;
                                    }
                                })
                                .filter(document -> {
                                    if (reason != null) {
                                        return document.getString("reason").equals(reason);
                                    } else {
                                        return true;
                                    }
                                })
                                .filterWhen(document -> userId
                                        .map(it -> document.getLong("userId").equals(it.asLong()))
                                        .defaultIfEmpty(true))
                                .filterWhen(document -> memberWhoWarnedId
                                        .map(it -> document.getLong("memberWhoWarnedId").equals(it.asLong()))
                                        .defaultIfEmpty(true))
                                .flatMap(mongoCollection::deleteOne)
                                .cast(Object.class)
                                .flatMap($1 -> /*guild.getMemberById(userId).flatMap(member -> */event.getChannel().flatMap(it -> it.createMessage("Successfully removed warn(s)!" /*"Successfully removed warn(s) for " + member.getDisplayName() + ".")*/)))
                                .switchIfEmpty(event.getChannel().flatMap(it -> it.createMessage("The user does not have a warn with that reason!")))
                                .then();

                    }
                })
                .switchIfEmpty(event.getChannel().flatMap(it -> it.createMessage("You must have at least 1 search query!")).then());
    }
}

File: src/main/kotlin/io/github/xf8b/xf8bot/commands/administration/RemoveWarnCommand.kt
/*
 * Copyright (c) 2020, 2021 xf8b.
 *
 * This file is part of xf8bot.
 *
 * xf8bot is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * xf8bot is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with xf8bot.  If not, see <https://www.gnu.org/licenses/>.
 */

package io.github.xf8b.xf8bot.commands.administration

import com.google.common.collect.ImmutableList
import discord4j.common.util.Snowflake
import io.github.xf8b.utils.tuples.and
import io.github.xf8b.xf8bot.api.commands.Command
import io.github.xf8b.xf8bot.api.commands.CommandFiredEvent
import io.github.xf8b.xf8bot.api.commands.flags.StringFlag
import io.github.xf8b.xf8bot.database.actions.delete.RemoveWarnAction
import io.github.xf8b.xf8bot.database.actions.find.FindWarnsAction
import io.github.xf8b.xf8bot.util.*
import io.r2dbc.spi.Result
import reactor.core.publisher.Mono
import reactor.kotlin.core.publisher.cast
import reactor.kotlin.core.publisher.toFlux
import reactor.kotlin.core.publisher.toMono
import reactor.util.function.Tuples
import java.util.*

class RemoveWarnCommand : Command(
    name = "\${prefix}removewarn",
    description = """
    Removes the specified member's warns with the warnId and reason provided.
    If the reason is all, all warns will be removed. The warnId is not needed.
    If the warnId is all, all warns with the same reason will be removed.
    """.trimIndent(),
    commandType = CommandType.ADMINISTRATION,
    aliases = ("\${prefix}removewarns" to "\${prefix}rmwarn" and "\${prefix}rmwarns").toImmutableList(),
    flags = ImmutableList.of(MEMBER, WARNER, REASON, WARN_ID),
    administratorLevelRequired = 1
) {
    override fun onCommandFired(event: CommandFiredEvent): Mono<Void> {
        val guildId = event.guildId.get()

        val memberIdMono = event[MEMBER].toMono().flatMap { member ->
            InputParsing.parseUserId(event.guild, member)
                .map(Long::toSnowflake)
                .switchIfEmpty(event.channel
                    .flatMap { it.createMessage("No member found! This may be caused by 2+ people having the same username or nickname.") }
                    .then() // yes i know, very hacky
                    .cast())
        }

        val warnerIdMono = event[WARNER].toMono().flatMap { member ->
            InputParsing.parseUserId(event.guild, member)
                .map(Long::toSnowflake)
                .switchIfEmpty(event.channel
                    .flatMap { it.createMessage("No member found! This may be caused by 2+ people having the same username or nickname.") }
                    .then() // yes i know, very hacky
                    .cast())
        }

        val reason = event[REASON]
        val warnId = event[WARN_ID]

        val warns = memberIdMono.flatMap { event.xf8bot.botDatabase.execute(FindWarnsAction(guildId, memberId = it)) }
            .switchIfEmpty(event.xf8bot.botDatabase.execute(FindWarnsAction(guildId)))
            .flatMapMany(List<Result>::toFlux)

        return Mono.zip(
            { array -> array.cast<Boolean>().any { present -> present } },
            Mono.just(warnId != null),
            warnerIdMono.flux().count().map(1L::equals),
            memberIdMono.flux().count().map(1L::equals),
            Mono.just(reason != null),
        ).filter { it } // filter if none are present
            .switchIfEmpty(event.channel
                .flatMap { it.createMessage("You must have at least 1 search query!") }
                .then()
                .cast())
            .flatMap {
                if (reason != null && reason.equals("all", ignoreCase = true)) {
                    memberIdMono.switchIfEmpty(event.channel
                        .flatMap { it.createMessage("Cannot remove all warns without a user!") }
                        .then()
                        .cast())
                        .flatMap { _ ->
                            warns.flatMap { result ->
                                result.map { row, _ ->
                                    event.xf8bot.botDatabase.execute(
                                        RemoveWarnAction(
                                            guildId = (row["guildId", Long.JAVA_TYPE] as Long).toSnowflake(),
                                            memberId = (row["memberId", Long.JAVA_TYPE] as Long).toSnowflake(),
                                            warnerId = (row["warnerId", Long.JAVA_TYPE] as Long).toSnowflake(),
                                            warnId = row["warnId", UUID::class.java],
                                            reason = row["reason", String::class.java]
                                        )
                                    )
                                }
                            }.then(event.channel
                                .flatMap { it.createMessage("Successfully removed all warns!") }
                                .then())
                        }
                } else {
                    Mono.zip(memberIdMono, warnerIdMono)
                        .defaultIfEmpty(Tuples.of(0L.toSnowflake(), 0L.toSnowflake()))
                        .flatMap { ids ->
                            val (memberId, warnerId) = ids.cast<Snowflake>().map { snowflake ->
                                if (snowflake.asLong() == 0L) null else snowflake
                            }

                            event.xf8bot.botDatabase.execute(
                                RemoveWarnAction(
                                    guildId,
                                    memberId,
                                    warnerId,
                                    warnId?.toUuid(),
                                    reason
                                )
                            )
                        }
                        .then(event.channel
                            .flatMap { it.createMessage("Successfully removed warn(s)!") }
                            .then())
                }
            }
    }

    companion object {
        private const val UUID_REGEX = "\\b[0-9a-f]{8}\\b-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-\\b[0-9a-f]{12}\\b"
        private val MEMBER = StringFlag(
            shortName = "m",
            longName = "member",
            required = false
        )
        private val WARNER = StringFlag(
            shortName = "w",
            longName = "warner",
            required = false
        )
        private val REASON = StringFlag(
            shortName = "r",
            longName = "reason",
            required = false
        )
        private val WARN_ID = StringFlag(
            shortName = "i",
            longName = "warnId",
            validityPredicate = { it matches UUID_REGEX.toRegex() },
            errorMessageFunction = { "The warn ID must be a UUID!" },
            required = false
        )
    }
}